### DISCUSSION TITLE
Question About `radam.py`

### DISCUSSION THREAD

**Discussion by giladturok** [geoopt/geoopt#216] [Q&A]
I'm trying to understand the implementation of Riemannian Adam in `radam.py` and am comparing it to the psuedo-code in the reference paper [Riemanian Adaptive Optimization Methods](https://arxiv.org/pdf/1810.00760.pdf).

#### Algorithm Definition
The paper contains the following algorithm for Riemannian Adam:

<img width="460" alt="Screen Shot 2023-04-02 at 11 21 57 AM" src="https://user-images.githubusercontent.com/36947659/229362416-9dbd51d0-e045-433e-9d62-a2f8fbde882f.png">

Note that in the definition of $v_t^i = \beta_2 v_{t-1}^i + (1 - \beta_2) ||g_t^i ||^2_{x_t^i}$ we observe that the gradient norm $||g_t^i ||_{x_t^i}$ is squared. 

#### Algorithm Implementation
This equation is defined in lines 92-94 of the `radam.py` file:

`exp_avg_sq.mul_(betas[1]).add_(manifold.component_inner(point, grad), alpha=1 - betas[1])`

#### Question

I was wondering where in this code is the gradient norm squared? I investigated `mainfold.component_inner` which seems to just take the inner product of `point` and `grad`. When lining up the variables in the paper and implementation, I could not find where this squared operation occurs.

Thus can someone please clarify where this squared operation is located? If the squaring occurs in `manifold.component_inner`, where/how does this happen? Thanks!

**Comment by giladturok** [comment]
Anyone have any insights? It would be much appreciated! 😄

**Comment by ferrine** [comment]
The inner product of gradient at a point is the squared norm already
